js/packages/mobile-wallet-adapter-protocol/android/src/main/java/com/solanamobile/mobilewalletadapter/reactnative/SolanaMobileWalletAdapterModule.kt

```diff
@@ -47,42 +46,37 @@ class SolanaMobileWalletAdapterModule(reactContext: ReactApplicationContext) :
         mutex.lock()
         Log.d(name, "startSession with config $config")
         try {
-            val semConnectedOrFailed = Semaphore(1, 1)
-            val semTerminated = Semaphore(1, 1)
             val uriPrefix = config?.getString("baseUri")?.let { Uri.parse(it) }
-            lateinit var localAssociation: LocalAssociationScenario
-            val scenarioCallbacks = object : Scenario.Callbacks {
-                override fun onScenarioReady(client: MobileWalletAdapterClient) {
-                    sessionState = SessionState(client, localAssociation, semTerminated)
-                    semConnectedOrFailed.release()
-                }
-
-                override fun onScenarioError() = semConnectedOrFailed.release()
-                override fun onScenarioComplete() = semConnectedOrFailed.release()
-                override fun onScenarioTeardownComplete() = semTerminated.release()
-            }
-            localAssociation = LocalAssociationScenario(
-                Looper.getMainLooper(),
-                ASSOCIATION_TIMEOUT_MS,
-                scenarioCallbacks,
-                uriPrefix
+            val localAssociation = LocalAssociationScenario(
+                CLIENT_TIMEOUT_MS,
+            )
+            val intent = LocalAssociationIntentCreator.createAssociationIntent(
+                uriPrefix,
+                localAssociation.port,
+                localAssociation.session
             )
-            val intent = localAssociation.createAssociationIntent()
             currentActivity?.startActivityForResult(intent, 0)
                 ?: throw NullPointerException("Could not find a current activity from which to launch a local association")
-            localAssociation.start()
-            withTimeout(ASSOCIATION_TIMEOUT_MS.toLong()) {
-                semConnectedOrFailed.acquire()
-            }
+            val client =
+                localAssociation.start().get(ASSOCIATION_TIMEOUT_MS.toLong(), TimeUnit.MILLISECONDS)
+            sessionState = SessionState(client, localAssociation)
             promise.resolve(true)
         } catch (e: ActivityNotFoundException) {
             Log.e(name, "Found no installed wallet that supports the mobile wallet protocol", e)
             cleanup()
             promise.reject("ERROR_WALLET_NOT_FOUND", e)
-        } catch (e: TimeoutCancellationException) {
+        } catch (e: TimeoutException) {
             Log.e(name, "Timed out waiting for local association to be ready", e)
             cleanup()
             promise.reject("Timed out waiting for local association to be ready", e)
+        } catch (e: InterruptedException) {
+            Log.w(name, "Interrupted while waiting for local association to be ready", e)
+            cleanup()
+            promise.reject(e)
+        } catch (e: ExecutionException) {
+            Log.e(name, "Failed establishing local association with wallet", e.cause)
+            cleanup()
+            promise.reject(e)
```

Comment:
Are these different or should one of them be eliminated now?

Reply:
These can still both be thrown.

`InterruptedException` is a generic Java exception for when `Thread.interrupt` is called, to cancel a normally blocking I/O method.

`ExecutionException` is the ugly exception output from `Future.get`; it wraps all exceptions other than the ones explicitly defined for `Future.get`. It makes sane exception handling with Futures harder than it needs to be :(